File: lib/ir/module.cpp
#include "evir/ir/module.hpp"
#include "evir/ir/comment.hpp"
#include "evir/ir/value/constant.hpp"
#include "evir/ir/value/reference.hpp"

using namespace evir;

#pragma region Constructors

Module::Module() {}

Module::Module(String name): name(name)
{
	add_metadata(new Metadata(Metadata::MD_MODULE_NAME, new MDStringValue(name)));
}

String Module::get_name()
{
	// metadata isn't set
	if(!has_metadata(Metadata::MD_MODULE_NAME)) return "";
	
	// metadata isn't a string
	Metadata* mdata = get_metadata(Metadata::MD_MODULE_NAME);
	if(!mdata->p_value->is_string()) return "";

	return ((MDStringValue*)mdata->p_value)->string;
}

#pragma endregion
#pragma region Metadata manipulation

bool Module::has_metadata(Metadata::Path path)
{
	// get it, and if it's null it isn't found
	return (bool)get_metadata(path);
}

bool Module::has_metadata(Metadata::BuiltinPropertyID id)
{
	// get it, and if it's null it isn't found
	return (bool)get_metadata(id);
}

void Module::add_metadata(Metadata* mdata)
{
	EVIR_ASSERT(!has_metadata(mdata->p_path), "Module already has metadata with similar path!");
	metadata.push_back(mdata);
}

void Module::add_metadata(Metadata::BuiltinPropertyID id, MDValue* value)
{
	EVIR_ASSERT(!has_metadata(id), "Module already has metadata with similar path!");
	add_metadata(new Metadata(id, value));
}

void Module::set_metadata(Metadata::Path path, MDValue* value)
{
	Metadata* mdata = get_metadata(path);
	EVIR_ASSERT(mdata, "Cannot set value of non-existent metadata property!");

	mdata->p_value = value;
}

void Module::set_metadata(Metadata::BuiltinPropertyID id, MDValue* value)
{
	Metadata* mdata = get_metadata(id);
	EVIR_ASSERT(mdata, "Cannot set value of non-existent metadata property!");

	mdata->p_value = value;
}

Metadata* Module::get_metadata(Metadata::Path path)
{
	// TODO?: improve?

	for(auto md : metadata) if(md->p_path == path) return md;
	return nullptr;
}

Metadata* Module::get_metadata(Metadata::BuiltinPropertyID id)
{
	// TODO?: improve?

	for(auto md : metadata) if(md->p_id == (Metadata::PropertyID)id) return md;
	return nullptr;
}

#pragma endregion
#pragma region User manipulation

Function* Module::get_function(String name)
{
	for(auto u : users)
		if(u->type->is_function_type() && u->get_name() == name)
			return (Function*)u;

	return nullptr;
}

Function* Module::get_or_insert_function(FunctionType* type, String name)
{
	Function* func = get_function(name);

	// not found
	if(!func)
	{
		func = new Function(type, name);
		users.push_back(func);
		return func;
	}

	// wrong type
	else if(*func->type != *type) return nullptr;

	// found
	else return func;
}

// Global* Module::get_or_insert_global(Type* type, String name)
// {
// 	for(auto u : users)
// 	{
// 		if(u->get_name() != name) continue;
// 		else if(*u->type != *type) return nullptr;
// 		else return (Global*)u;
// 	}
//
// 	Global* glob = new Global(type, name);
// 	users.push_back(glob);
// 	return glob;
// }

#pragma endregion
#pragma region IR generation

String Module::generate_metadata_ir(bool before_contents)
{
	#define STREAM(name) SStream s_##name = SStream()

	STREAM(module);
	STREAM(module_source);
	STREAM(module_producer);
	STREAM(target);
	STREAM(debug);
	STREAM(custom);

	#undef STREAM
	#define HANDLE_RANGE(do, name) if(do && md->p_id > Metadata::_MD_start_##name && \
								  md->p_id < Metadata::_MD_end_##name) \
							   { s_##name << md->generate_ir() << endl; continue; }

	// gather metadata ir in seperate streams
	for(auto md : metadata)
	{
		HANDLE_RANGE(before_contents, module);
		HANDLE_RANGE(before_contents, module_source);
		HANDLE_RANGE(before_contents, module_producer);

		// debug/typenames is on a new line
		if(!before_contents && md->p_id == Metadata::_MD_DEBUG_TYPENAMES) s_debug << endl;

		HANDLE_RANGE(!before_contents, target);
		HANDLE_RANGE(!before_contents, debug);

		// custom metadata follows at the end
		if(!before_contents && md->p_id == Metadata::_MD_CUSTOM_)
			s_custom << md->generate_ir() << endl;	
	}

	#undef HANDLE_RANGE
	#define EMPTY(name) (s_##name.str().empty())
	#define EMIT_STREAM(name) if(!EMPTY(name)) stream << s_##name.str() << endl

	// concat all streams in correct order
	SStream stream = SStream();
	if(before_contents)
	{
		// module/...
		if(!(EMPTY(module) && EMPTY(module_source) && EMPTY(module_producer)))
			stream << generate_ir_comment("module metadata", true);
		EMIT_STREAM(module);
		EMIT_STREAM(module_source);
		EMIT_STREAM(module_producer);
	}
	else
	{
		// target/...
		if(!EMPTY(target)) stream << generate_ir_comment("target metadata", true);
		EMIT_STREAM(target);

		// debug/...
		if(!EMPTY(debug)) stream << generate_ir_comment("debug info metadata", true);
		EMIT_STREAM(debug);

		// ...
		EMIT_STREAM(custom);
	}

	#undef EMPTY
	#undef EMIT_STREAM

	// clean up and return
	String ir = stream.str();
	if(*ir.end() == endl) ir.erase(ir.end() - 1);
	return ir; 
}

String Module::generate_ir()
{
	SStream stream = SStream();

	// generate metata ir before contents
	stream << generate_metadata_ir(true);
	
	// generate users' ir
	stream << generate_ir_comment("module contents", true);
	stream << endl;
	for(auto user : users) stream << user->generate_ir() << endl;

	// generate metata ir after contents
	stream << generate_metadata_ir(false);

	// generate credit comments
	stream << generate_ir_comment("module generated with:");
	#ifdef DEBUG
	stream << generate_ir_comment("\t" EVIR_LIB_NAME_INTERNAL " version " EVIR_LIB_VERSION " dev");
	#else
	stream << generate_ir_comment("\t" EVIR_LIB_NAME_INTERNAL " version " EVIR_LIB_VERSION);
	#endif
	stream << generate_ir_comment("\tby Sjoerd Vermeulen");
	stream << generate_ir_comment("\tMIT license (2022)");

	// return string (minus last newline)
	return stream.str().erase(stream.str().length() - 1);
}

#pragma endregion

File: lib/ir/module.rs
//===--- ir/module.rs ------------------ Rust ---===
// This header is part of the EvIr library
// MIT License - Sjoerd Vermeulen (2022)
// For more info see https://github.com/SjVer/EvIr
//===--------------------------------------------===

use crate::{
	ENDL, LIB_NAME_INTERNAL, LIB_VERSION,
	__evir_reset_tmp_name_count,
	ir::{
		IR, generate_ir_comment,
		metadata::*,
		Function, FUNCTION_TMPNAMEGETTER,
		FunctionType,
	},
};

/// A single stand-alone IR module. \
/// Has the following metadata by default:
/// - [!module/name][`BuiltinMDProp::ModuleName`]
/// - [!debug/generate][`BuiltinMDProp::DebugGenerate`]
pub struct Module {
	metadata: Vec<Metadata>,
	functions: Vec<Function>,
}

// misc stuff
impl Module {
	pub fn new(name: impl ToString) -> Self {
		let mut this = Self {
			metadata: vec![],
			functions: vec![],
		};

		this.set_metadata(Metadata::new(BuiltinMDProp::ModuleName, name.to_string()));
		// this.set_metadata(Metadata::new(BuiltinMDProp::DebugGenerate, ConstantInt::new(0)));

		this
	}

	pub fn get_name(&mut self) -> Option<String> {
		match self.get_metadata(BuiltinMDProp::ModuleName) {
			Some(md) => match &md.value {
				MDValue::String(s) => Some(s.clone()),
				_ => None,
			}
			None => None,
		}
	}
}

// metadata stuff
impl Module {
	/// Attempts to get (a mutable reference to) the 
	/// metadata value at the given [path][`MDPath`].
	pub fn get_metadata(&mut self, path: impl ToMDPath) -> Option<&mut Metadata> {
		let path = path.to_mdpath();

		for md in &mut self.metadata {
			if md.path == path { return Some(md); }
		}

		None
	}

	/// Checks if the module has metadata set at
	/// the given path.
	pub fn has_metadata(&mut self, path: impl ToMDPath) -> bool {
		self.get_metadata(path).is_some()
	}

	/// Sets a metadata value at the given path.
	/// If a value was already set it is overwritten.
	pub fn set_metadata(&mut self, md: Metadata) {
		match self.get_metadata(md.path.clone()) {
			Some(old) => old.value = md.value,
			None => self.metadata.push(md)
		}
	}
}

// state stuff
impl Module {
	/// Returns the function if it exists, otherwise None.
	pub fn get_function(&mut self, name: impl ToString) -> Option<&mut Function> {
		let name = name.to_string();
		
		for f in &mut self.functions {
			if let Some(n) = f.get_name() {
				if n == name {
					return Some(f);
				}
			}
		}

		None
	}

	/// Returns the function if it exists, or a new function
	/// if it doesn't. None is returned if the function exists
	/// but with a different type, or if a different kind of
	/// user with the same name exists.
	pub fn get_or_create_function(&mut self, name: impl ToString, ftype: FunctionType) -> Option<&mut Function> {
		if let Some(f) = self.get_function(name.to_string()) {
			// if f.get_type() == &ftype { Some(f) }
			// else { None }
			None
		} else {
			// Other user with same name? None
			// else:
			self.functions.push(Function::new(Some(name.to_string()), ftype));
			self.functions.last_mut()
		}
	}

	/// Creates a new nameless function. (The name of a function 
	/// is optional. It can be set using [set_name()][`Function::set_name`])
	pub fn create_nameless_function(&mut self, ftype: FunctionType) -> &mut Function {
		self.functions.push(Function::new(None, ftype));
		self.functions.last_mut().unwrap()
	}
}

// ir generation
impl Module {
	fn generate_metadata_ir(&mut self, before_contents: bool) -> IR {
		// ===== first gather ir for each part seperatly =====
		
		let mut m  = IR::new(); // !module/...
		let mut ms = IR::new(); // !module/source/...
		let mut mp = IR::new(); // !module/producer/...
		let mut t  = IR::new(); // !target/...
		let mut d  = IR::new(); // !debug/...
		let mut c  = IR::new(); // !...

		macro_rules! do_prop {
			($what:ident $path:ident) => {
				if let Some(md) = self.get_metadata(BuiltinMDProp::$path) {
					$what += &md.generate_ir();
				}
			};
		}
		macro_rules! do_cust {
			($what:ident $path:ident) => {
				for md in &self.metadata {
					if let MDPath::Custom(s, ..) = &md.path {
						if s == &CustomMDProp::$path {
							$what += &md.generate_ir();
						}
					}
				} 
			};
		}

		do_prop!(m ModuleName);
		do_prop!(m ModuleEntrypoint);

		do_prop!(ms ModuleSourceFilename);
		do_prop!(ms ModuleSourceDirectory);
		do_prop!(ms ModuleSourceLanguage);
		do_cust!(ms ModuleSource);

		do_prop!(mp ModuleProducerName);
		do_prop!(mp ModuleProducerVersion);
		do_prop!(mp ModuleProducerType);
		do_cust!(mp ModuleProducer);

		do_cust!(m Module);

		do_prop!(t TargetTriple);
		do_prop!(t TargetCPU);
		do_prop!(t TargetDatalayout);
		do_prop!(t TargetOptimization);
		do_cust!(t Target);

		do_prop!(d DebugGenerate);
		do_prop!(d DebugIncludesource);
		do_prop!(d DebugSourcelocation);
		do_prop!(d DebugSourcechecksum);
		do_prop!(d DebugDwarfversion);
		do_prop!(d DebugTypenames);
		do_cust!(d Debug);

		do_cust!(c Custom);
		
		// ===== then generate and join ir if need be =====
		
		let mut ir = IR::new();

		macro_rules! emit {
			($ir:expr) => {
				if !$ir.is_empty() {
					ir += &$ir;
					ir.push_str(ENDL);
				}
			};
		}

		if before_contents {
			if !m.is_empty() || !ms.is_empty() || !mp.is_empty() {
				ir += &generate_ir_comment("module metadata", true);
			}

			emit!(m);
			emit!(ms);
			emit!(mp);
		} else {
			if !t.is_empty() { ir += &generate_ir_comment("target metadata", true); }
			emit!(t);

			if !d.is_empty() { ir += &generate_ir_comment("debug info metadata", true); }
			emit!(d);
			
			emit!(c);
		}

		// return it
		ir
	}

	/// Generates the complete IR for the module.
	pub fn generate_ir(&mut self) -> IR {
		let mut ir = IR::new();

		// generate metadata ir first
		ir += &self.generate_metadata_ir(true);
		
		// generate users' ir
		ir += &generate_ir_comment("module contents", true);
		ir.push('\n');
		
		// functions
		__evir_reset_tmp_name_count!(FUNCTION_TMPNAMEGETTER);
		for u in &mut self.functions {
			ir += &u.generate_ir();
			ir.push('\n');
		}
		
		// generate metadata after contents
		ir += &self.generate_metadata_ir(false);

		// credits comment
		ir += &generate_ir_comment("module generated with:", false);
		ir += &generate_ir_comment(format!("\t{} version {}", LIB_NAME_INTERNAL, LIB_VERSION), false);
		ir += &generate_ir_comment("\tby Sjoerd Vermeulen", false);
		ir += &generate_ir_comment("\tMIT license (2022)", false);

		// return the ir
		ir.trim().to_string()
	}
}